src/Common/src/Interop/OSX/Interop.libproc.cs

```diff
@@ -0,0 +1,523 @@
+        /// </summary>
+        /// <returns>Returns a list of open File Descriptors for this process</returns>
+        /// <remarks>
+        /// This function doesn't use the helper since it seems to allow passing NULL
+        /// values in to the buffer and length parameters to get back an estimation 
+        /// of how much data we will need to allocate; the other flavors don't seem
+        /// to support doing that.
+        /// </remarks>
+        internal static List<proc_fdinfo> GetFileDescriptorsForPid(int pid)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid");
+            }
+
+            int size = 0;
+            IntPtr buffer = IntPtr.Zero;
+            
+            // Query for an estimation about the size of the buffer we will need. This seems
+            // to add some padding from the real number, so we don't need to do that
+            int result = proc_pidinfo(pid, PROC_PIDLISTFDS, 0, buffer, size);
+            if (result <= 0)
+            {
+                throw new System.ComponentModel.Win32Exception();
+            }
+
+            // Allocate our unmanaged memory and call again to get the descriptors
+            size = result;
+            buffer = Marshal.AllocCoTaskMem(size);
+            result = proc_pidinfo(pid, PROC_PIDLISTFDS, 0, buffer, size);
+            if (result <= 0)
+            {
+                throw new System.ComponentModel.Win32Exception();
+            }
+
+            List<proc_fdinfo> fds = new List<proc_fdinfo>();
+
+            // Copy the file descriptors over
+            for (int i = 0; i < (result / PROC_PIDLISTFD_SIZE); i++)
+            {
+                fds.Add(Marshal.PtrToStructure<proc_fdinfo>(buffer + (i * PROC_PIDLISTFD_SIZE)));
+            }
+
+            // Free our unmanaged memory
+            Marshal.FreeCoTaskMem(buffer);
+
+            return fds;
+        }
+
+        private static bool internal_proc_pidinfo_helper<T>(int pid, int flavor, ulong args, out T info)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid", SR.NegativePidNotSupported);
+            }
+
+            bool success = false;
+
+            // Allocate our unmanaged buffer for the task info
+            info = default(T);
+            int size = Marshal.SizeOf<T>();
+            IntPtr buffer = Marshal.AllocCoTaskMem(size);
+
+            // Get the process information for the given PID
+            int result = proc_pidinfo(pid, flavor, args, buffer, size);
+            if (result == size)
+            {
+                // Copy the struct from unmanaged memory to managed memory
+                info = Marshal.PtrToStructure<T>(buffer);
+                success = true;
+            }
+
+            // Free our unmanaged buffer
+            Marshal.FreeCoTaskMem(buffer);
```

Comment:
Same comment about a finally being needed (and about HGlobal instead of CoTaskMem... I'll stop commenting on these).